src/main/java/net/neoforged/neoforge/transfer/fluids/FluidResource.java

```diff
@@ -0,0 +1,244 @@
+     * @param patch the patch to apply
+     * @return the new resource
+     */
+    public FluidResource applyPatch(DataComponentPatch patch) {
+        FluidStack stack = innerStack.copy();
+        stack.applyComponents(patch);
+        return new FluidResource(stack);
+    }
+
+    /**
+     * Returns a copy of this resource with the set data component.
+     * @param type the type of data component
+     * @param data the data to set
+     * @return the new resource
+     * @param <D> the type of data component
+     */
+    public <D> FluidResource set(DataComponentType<D> type, D data) {
+        FluidStack stack = innerStack.copy();
+        stack.set(type, data);
+        return new FluidResource(stack);
+    }
+
+    /**
+     * Returns a copy of this resource with the data component removed.
+     * @param type the type of data component
+     * @return the new resource
+     */
+    public FluidResource remove(DataComponentType<?> type) {
+        FluidStack stack = innerStack.copy();
+        stack.remove(type);
+        return new FluidResource(stack);
+    }
+
+    /**
+     * @return the fluid of this resource
+     */
+    public Fluid getFluid() {
+        return innerStack.getFluid();
+    }
+
+    /**
+     * @return the fluid holder of this resource
+     */
+    public Holder<Fluid> getFluidHolder() {
+        return innerStack.getFluidHolder();
+    }
+
+    /**
+     * @return the fluid type of this resource
+     */
+    public FluidType getFluidType() {
+        return innerStack.getFluidType();
+    }
+
+    @Override
+    public DataComponentMap getComponents() {
+        return innerStack.getComponents();
+    }
+
+    public DataComponentPatch getComponentsPatch() {
+        return innerStack.getComponentsPatch();
+    }
+
+    public boolean matches(FluidStack stack) {
+        return FluidStack.isSameFluidSameComponents(stack, innerStack);
+    }
+
+    public FluidStack toStack(int amount) {
+        return this.innerStack.copyWithAmount(amount);
+    }
+
+    public FluidStack toStack() {
+        return toStack(FluidType.BUCKET_VOLUME);
+    }
+
+    public boolean is(TagKey<Fluid> tag) {
+        return innerStack.is(tag);
+    }
+
+    public boolean is(Fluid fluid) {
+        return innerStack.is(fluid);
+    }
+
+    public boolean is(Predicate<Holder<Fluid>> predicate) {
+        return innerStack.is(predicate);
+    }
+
+    public boolean is(Holder<Fluid> holder) {
+        return innerStack.is(holder);
+    }
+
+    public boolean is(HolderSet<Fluid> holders) {
+        return innerStack.is(holders);
+    }
+
+    public boolean is(FluidType fluidType) {
+        return innerStack.is(fluidType);
+    }
+
+    public boolean isVaporizedOnPlacement(Level level, BlockPos pos) {
+        return innerStack.getFluidType().isVaporizedOnPlacement(level, pos, innerStack);
+    }
+
+    public void onVaporize(Player player, Level level, BlockPos pos) {
```

Comment:
So I understand these are to not bypass immutability of fluidstack, but I feel the use cases for each call is going to be a nightmare to support. getTemperature or viscosity for example would be things I'd want, but then need the API to expose.